src/fetchers/tarball-fetcher.js

```diff
@@ -17,6 +17,8 @@ const stream = require('stream');
 const gunzip = require('gunzip-maybe');
 const invariant = require('invariant');
 
+const RE_URL_NAME_MATCH = /\/(?:(@[^/]+)\/)?[^/]+\/-\/([^/]+)$/;
```

Comment:
I think this will fail for some packages.  For example, looking at some sample resolved urls in a random local yarn.lock, it appears that _most_ would match your regex, such as ones of the form

https://publish.internal.site:443/artifactory/api/npm/external-npm/abab/-/abab-1.0.3.tgz
or
https://publish.internal.site:443/artifactory/api/npm/external-npm/@types/react/-/react-15.0.38.tgz

However, there are also some of the form

https://artifactory.internal.site:443/artifactory/api/npm/internal-npm-release/@team/package/-/@team/package-1.2.3.tgz

It looks like all internally team-scoped packages appear to have an extra "/@team/" between the "/-/" and the tarballBasename.

Reply:
That's quite weird and it was also what I feared. We can tune the regex to handle these too but ideally, we should not be extracting the scope information from the URL anyway.

Since changing the names of these files would invalidate any existing offline mirror, we need to find a migration path when doing this in a better way.

Reply:
I've pushed a new change but I also need to extend and refactor the tests to cover this new scenario. Feel free to add them since I can't look at this patch until tomorrow.